### PR TITLE
Assignment 4 Alvaro

### PR DESCRIPTION
**Pull request by alvaro-dda**
# Main considerations
The first part, until Non-Linear Models is the same as the improved version of Assignment 3.

The models used are:
- Bagging
- Random Forest
- Decision Tree
- Gradient Boosting Trees (XGBoost)

### Overall Performance

The model used for the second part of the assignment is the XGBoost with 200 trees and a learning rate of 0.01 which managed to beat the log reg model with L1 regularisation. Although the Bagging model obtained the same scores for the desired threshold, I chose the XGBoost one because it was faster to train and it is better for other FPR thresholds. 

## Final Results for FPR from 0 to 0.05%
Screenshot 2023-10-22 at 14 54 38

### CODE REVIEW

File: src/module_4/assign_4_model_predict.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    for user, prediction in zip(user_predictions_df.index, predictions):
+        json_data[str(user)] = int(prediction)
+    
+    return {
+    "statusCode": "200",
+    "body": json.dumps(
+        { "prediction": json_data})
+        }
+
+event=dict()
```
Comment (guillerbf) [inline src/module_4/assign_4_model_predict.py:23]: This test should not be here, you can create a `test_` file even if it's not a util test but some kind of integration test where you just want to validate that it runs and generates the desired output.

File: src/module_4/assign_4_model_predict.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+def handler_predict(event,model_path):
+    user_predictions_df = pd.DataFrame.from_dict(json.loads(event["users"]),orient='index')
+    
+    with open(model_path, 'rb') as model_file:
+        model = pickle.load(model_file)
+    
+    predictions = model.predict(user_predictions_df)
+
+    json_data = {}
+    for user, prediction in zip(user_predictions_df.index, predictions):
```
Comment (guillerbf) [inline src/module_4/assign_4_model_predict.py:14]: or even nicer as a list comprehension:

`json_data = {str(user): int(prediction) for user, prediction in zip(user_predictions_df.index, predictions)}`

File: src/module_4/assign_4_model_predict.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import pickle
+import pandas as pd
+import json
+
+def handler_predict(event,model_path):
```
Comment (guillerbf) [inline src/module_4/assign_4_model_predict.py:5]: Add types!

Add docstring defining what format you expect in `event`. You don't need it all and it's kinda self documented in the "test" that you added in the same file but you could add some simple docstring like:

```
def handler_predict(event: dict, model_path: str) -> dict:
"""
  event: {
      'user_id_1': {
           feature_1: xxx,
           feature_2: yyy,
           ...
      },
      'user_id_2': {
           ...
      }
 }
 """

File: src/module_4/assign_4_model_training.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import pickle
+from datetime import datetime
+
+def load_data(path: str) -> pd.DataFrame:
+    """
+    Loads and cleans the data
+    """
+    df = pd.read_csv(path)
+
+    #Filtering for orders with at least 5 purchases
```
Comment (guillerbf) [inline src/module_4/assign_4_model_training.py:16]: This should go to a function to test it. 

Ideally, this could be reusable in notebook. Always trying to create single source of truth to avoid discrepancies.

File: src/module_4/assign_4_model_training.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    df_orders = df.groupby(by='order_id')['outcome'].sum()
+    valid_orders = df_orders[df_orders>=5].index.to_list() 
+    df = df[df['order_id'].isin(valid_orders)]
+
+    #We remove the ID Columns and Categorical
+    feature_columns = ['user_order_seq',
```
Comment (guillerbf) [inline src/module_4/assign_4_model_training.py:22]: This could be a global variable, `FEATURE_COLUMNS` at the top of the script or even better in a `config` file outside the script. The reason is so we can reuse the same features in train and predict.

File: src/module_4/assign_4_model_predict.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    user_predictions_df = pd.DataFrame.from_dict(json.loads(event["users"]),orient='index')
+    
+    with open(model_path, 'rb') as model_file:
+        model = pickle.load(model_file)
+    
+    predictions = model.predict(user_predictions_df)
```
Comment (guillerbf) [inline src/module_4/assign_4_model_predict.py:11]: How do you know the same features you have used in train are present here?

File: src/module_4/assign_4_model_training.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    return df['outcome'], df[feature_columns]
+
+def handler_fit(event:dict) -> dict:
+    '''
+    Trains a Gradient Boosting Decision Tree model based on the parameters specified in event.
+    Returns a json with the path of the saved model.
+    '''
+
+    if not isinstance(event["model_parametrisation"], dict):
+        raise ValueError("Parameter Grid is not a Dictionary.")
```
Comment (guillerbf) [inline src/module_4/assign_4_model_training.py:40]: Nice

File: src/module_4/assign_4_model_training.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+   
+    model_parametrisation = event["model_parametrisation"]
+
+    train_data_y, train_data_x = load_data('src/module_4/Datasets_Zrive/feature_frame.csv')
```
Comment (guillerbf) [inline src/module_4/assign_4_model_training.py:44]: data_path should be a global variable or ideally a parameter in `config` file

File: src/module_4/assign_4_model_training.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    
+    pipeline = Pipeline([
+        ('clf', XGBClassifier(**model_parametrisation))])
+    
+    model = pipeline.fit(train_data_x, train_data_y)
+    
+    today_date = datetime.now().strftime('%Y_%m_%d')
```
Comment (guillerbf) [inline src/module_4/assign_4_model_training.py:51]: Line 51 to 55 can be taken out to a function for testability, reusability and cleanliness.

File: src/module_4/assign_4_model_training.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    model_path = f'src/module_4/push_{today_date}.pkl'
+
+    with open(model_path, 'wb') as model_file:
+        pickle.dump(model, model_file)
+    
+    return {
+    "statusCode": "200",
+    "body": json.dumps(
+    {"model_path": [model_path]})}
```
Comment (guillerbf) [inline src/module_4/assign_4_model_training.py:60]: Why `[model_path]` rather than `model_path`?

File: src/module_4/assign_4.md [hunk 11]
```diff
@@ -0,0 +1,824 @@
...
+           'avg_days_to_buy_variant_id', 'std_days_to_buy_variant_id',
+           'days_since_purchase_product_type', 'avg_days_to_buy_product_type',
+           'std_days_to_buy_product_type'],
+          dtype='object')
+
+
+
+
+```python
+#Filtering for Valid Orders
```
Comment (guillerbf) [inline src/module_4/assign_4.md:45]: Wrap it into a function

File: src/module_4/assign_4.md [hunk 12]
```diff
@@ -0,0 +1,824 @@
...
+</table>
+<p>2163953 rows × 27 columns</p>
+</div>
+
+
+
+
+```python
+#Removing ID Variables and adding new features
+df = df.drop(['created_at','user_id','variant_id','order_id'], axis=1)
```
Comment (guillerbf) [inline src/module_4/assign_4.md:371]: Why do you need dropping them if you are doing explicit select of features down the line? It's typically nice to keep ID features just in case you want to do plots or analysis.

File: src/module_4/assign_4.md [hunk 13]
```diff
@@ -0,0 +1,824 @@
...
+
+```python
+#Linear Logistic Regression Models
+grid_search_l1 = get_reg_model('l1', df_train.loc[:,df_numeric_features])
+warnings.filterwarnings('ignore')
+```
+
+
+```python
+#FREQUENCY ENCODING of Categorical Variables for train and validation dataset
```
Comment (guillerbf) [inline src/module_4/assign_4.md:496]: Ideally wrap it into a function that can be tested even if you won't test it for the notebook (although you could)

File: src/module_4/assign_4.md [hunk 14]
```diff
@@ -0,0 +1,824 @@
...
+
+
+
+```python
+from sklearn.ensemble import RandomForestClassifier
+
+#Random Forest Model
+def get_model_forest(features: pd.DataFrame) -> GridSearchCV:    
+    param_grid = {
+        'clf__n_estimators': [30],
```
Comment (guillerbf) [inline src/module_4/assign_4.md:594]: knowing that the bigger the forest the better, I'd challenge why only 30 trees.

File: src/module_4/assign_4.md [hunk 15]
```diff
@@ -0,0 +1,824 @@
...
+draw_roc_prc(df_val['outcome'], list_val_probs, labels)
+```
+
+
+    
+![png](assign_4_files/assign_4_26_0.png)
+    
+
+
+Clearly both Bagging and Random Forest Classifier's are both underperforming on the validation dataset because of the small number of trees used (really slow to train). I'm going to try to fix the parameters and increase the number of trees to see if they improve.
```
Comment (guillerbf) [inline src/module_4/assign_4.md:712]: Good

File: src/module_4/assign_4.md [hunk 16]
```diff
@@ -0,0 +1,824 @@
...
+      if is_sparse(data):
+    /Users/alvarodda/Library/Caches/pypoetry/virtualenvs/zrive-ds-GYwyIoqR-py3.11/lib/python3.11/site-packages/xgboost/data.py:520: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
+      if is_sparse(data):
+    /Users/alvarodda/Library/Caches/pypoetry/virtualenvs/zrive-ds-GYwyIoqR-py3.11/lib/python3.11/site-packages/xgboost/data.py:520: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
+      if is_sparse(data):
+
+
+
+```python
+#Checking it for ROC from 0 to 0.05%
```
Comment (guillerbf) [inline src/module_4/assign_4.md:786]: Nice idea

**Review by guillerbf — COMMENTED** [review]
Good stuff! Few comments inline, most having to do with going from spaguetti code to modular code that can be tested, maintained and reused.

In `model_predict` you need to be careful with input data assumptions. You would benefit of having some shared function that selects features for both train and predict

BTW: I can't see the images in the markdown, you may want to attach them to the PR to.
